src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -411,6 +447,149 @@ void ReduceAxesCompute(const nnvm::NodeAttrs& attrs,
   ReduceAxesComputeImpl<xpu, reducer, normalize>(attrs, ctx, inputs, req, outputs, small);
 }
 
+template <int req, int axis>
+struct SumCsrKernel;
+
+template <int req>
+struct SumCsrKernel<req, 0> {
+  template <typename RType, typename IType, typename DType>
+  MSHADOW_XINLINE static void Map(int j, DType* out_data,
+                                  const RType* in_indptr, const IType* in_idx,
+                                  const DType* in_data,
+                                  const int64_t num_rows) {
+    DType sum, residual;
+    mshadow::red::sum::SetInitValue(sum, residual);
+    const IType jval = static_cast<IType>(j);
+    for (RType i = 0; i < num_rows; ++i) {
+      if (in_indptr[i] >= in_indptr[i + 1]) continue;
+      if ((in_idx[in_indptr[i]] <= jval) &&
+          (in_idx[in_indptr[i + 1] - 1] >= jval)) {
+        // Do binary search for j between in_idx[in_indptr[i]] and
+        // in_idx[in_indptr[i+1]]
+        // The assumption here is in_idx for each row is sorted
+        IType start = in_indptr[i];
+        IType end = in_indptr[i + 1] - 1;
```

Comment:
Can you please cache the in_indptr[i + 1] value? It is used a lot an can introduct a lot of superfluous instructions.

Reply:
fixed